examples/albef/finetune_retrieval.py

```diff
@@ -0,0 +1,304 @@
+from examples.albef.data.retrieval_datamodule import RetrievalDataModule
+from examples.albef.model import albef_model_for_retrieval
+from examples.albef.utils import add_weight_decay
+from torch.optim import AdamW
+from torch.optim.lr_scheduler import CosineAnnealingWarmRestarts
+
+
+def train(model, datamodule, args, checkpoint_root, device):
+    model.train()
+
+    model_without_ddp = (
+        model.module if dist_utils.is_dist_avail_and_initialized() else model
+    )
+
+    optimizer_params = add_weight_decay(model, args["weight_decay"])
+    optimizer = AdamW(optimizer_params, lr=args["lr"])
+    scheduler = CosineAnnealingWarmRestarts(
+        optimizer, T_0=args["max_epochs"], eta_min=args["min_lr"]
+    )
+
+    step_size = args["step_size"]
+    warmup_steps = args["warmup_steps"]
+    warmup_iterations = warmup_steps * step_size
+
+    data_loader = datamodule.train_dataloader(
+        is_distributed=dist_utils.is_dist_avail_and_initialized(),
+        num_tasks=dist_utils.get_world_size(),
+        global_rank=dist_utils.get_rank(),
+    )
+
+    dataset_length = len(datamodule.train_dataset)
+    start_time = time.time()
+    log_every_n_steps = 100
+
+    for epoch in range(args["max_epochs"]):
+        if epoch > 0:
+            scheduler.step(epoch + warmup_steps)
+
+        for batch, (image, text, text_atts, idx) in enumerate(data_loader):
+            if epoch > 0:
+                alpha = args["alpha"]
+            else:
+                alpha = args["alpha"] * min(1, batch / dataset_length)
+
+            image = image.to(device, non_blocking=True)
+            text = text.to(device)
+            text_atts = text_atts.to(device)
+            idx = idx.to(device, non_blocking=True)
+            loss = model(image, text, text_atts, idx, alpha, is_train=True)
+
+            optimizer.zero_grad()
+            loss.backward()
+            optimizer.step()
+
+            if epoch == 0 and batch % step_size == 0 and batch <= warmup_iterations:
+                scheduler.step(batch // step_size)
+
+            if batch % log_every_n_steps == 0:
+                total_time = time.time() - start_time
+                time_str = "time {},".format(
+                    datetime.timedelta(seconds=int(total_time))
+                )
+                epoch_str = "epoch {}/{},".format(epoch, args["max_epochs"])
+                batch_str = "batch {}/{},".format(batch, len(data_loader))
+                loss_str = "loss {}".format(loss.item())
+                print(time_str, epoch_str, batch_str, loss_str)
+
+        save_obj = {
+            "model": model_without_ddp.state_dict(),
+            "optimizer": optimizer.state_dict(),
+            "lr_scheduler": scheduler.state_dict(),
+            "epoch": epoch,
+        }
+        torch.save(
+            save_obj,
+            os.path.join(checkpoint_root, "retrieval_checkpoint_%02d.pth" % epoch),
+        )
+
+
+@torch.no_grad()
+def evaluation(model, datamodule, k, device):
```

Comment:
Maybe split this up into multiple methods for handling text features, image features, and text_to_image/image_to_text_scores? As is it is a bit hard to follow